File: Problems/No.641_Design_Circular_Dequeue/design_circular_dequeue.py
class MyCircularDeque(object):

    def __init__(self, k):
        """
        :type k: int
        """
        self.MAX_size = k
        self.curr_size = 0
        self.DQueue = [0] * self.MAX_size
        self.front = 0
        

    def insertFront(self, value):
        """
        :type value: int
        :rtype: bool
        """
        if self.isFull():
            return False
        else:
            self.front -= 1
            self.DQueue[self.front] = value
            self.curr_size += 1
            return True
        

    def insertLast(self, value):
        """
        :type value: int
        :rtype: bool
        """
        if self.isFull():
            return False
        else:
            self.DQueue[self.front + self.curr_size] = value
            self.curr_size += 1
            return True


    def deleteFront(self):
        """
        :rtype: bool
        """
        if self.isEmpty():
            return False
        else:
            self.front += 1
            self.curr_size -= 1
            return True
        

    def deleteLast(self):
        """
        :rtype: bool
        """
        if self.isEmpty():
            return False
        else:
            self.curr_size -= 1
            return True
        

    def getFront(self):
        """
        :rtype: int
        """
        return self.DQueue[self.front] if not self.isEmpty() else -1
        

    def getRear(self):
        """
        :rtype: int
        """
        return self.DQueue[self.front + self.curr_size - 1] if not self.isEmpty() else -1
        

    def isEmpty(self):
        """
        :rtype: bool
        """
        return (self.curr_size == 0)
        

    def isFull(self):
        """
        :rtype: bool
        """
        return (self.curr_size == self.MAX_size)
        


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()

File: Problems/No.641_Design_Circular_Dequeue/design_circular_dequeue.cpp
class MyCircularDeque {

private:
    std::vector<int> DQueue;
    int MAX_size, curr_size;
    int front, rear;

public:
    MyCircularDeque(int k) {
        
        MAX_size = k;
        curr_size = 0;
        DQueue.resize(MAX_size);
        front = 0;

    }
    
    bool insertFront(int value) {
        
        if (isFull()) return false;
        else {

            front = (front - 1 + MAX_size)%MAX_size;
            DQueue[front] = value;
            curr_size++;
            return true;

        }

    }
    
    bool insertLast(int value) {
        
        if (isFull()) return false;
        else {

            DQueue[(front + curr_size)%MAX_size] = value;
            curr_size++;
            return true;

        }

    }
    
    bool deleteFront() {
        
        if (isEmpty()) return false;
        else {

            front = (front + 1)%MAX_size;
            curr_size--;
            return true;

        }

    }
    
    bool deleteLast() {
        
        if (isEmpty()) return false;
        else {

            curr_size--;
            return true;

        }

    }
    
    int getFront() {return isEmpty() ? -1 : DQueue[front];}
    
    int getRear() {return isEmpty() ? -1 : DQueue[(front + curr_size - 1)%MAX_size];}
    
    bool isEmpty() {return (curr_size == 0);}
    
    bool isFull() {return (curr_size == MAX_size);}
};

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque* obj = new MyCircularDeque(k);
 * bool param_1 = obj->insertFront(value);
 * bool param_2 = obj->insertLast(value);
 * bool param_3 = obj->deleteFront();
 * bool param_4 = obj->deleteLast();
 * int param_5 = obj->getFront();
 * int param_6 = obj->getRear();
 * bool param_7 = obj->isEmpty();
 * bool param_8 = obj->isFull();
 */
